File: DCC2024_Artur/February/Day3_Partition_Array_for_Maximum_Sum.cpp
/*
 https://leetcode.com/problems/partition-array-for-maximum-sum/description/?envType=daily-question&envId=2024-02-03
*/

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int maxSumAfterPartitioning(vector<int>& arr, int k)
    {
        vector<int> dp = vector<int>(arr.size(), 0);
        dp[0] = arr[0];

        int currMax = arr[0];
        for(int i = 1; i < k; ++i)
        {
            currMax = max(currMax, arr[i]);
            dp[i] = currMax * (i + 1);
        }

        for(int i = k; i < arr.size(); ++i)
        {
            currMax = arr[i];
            for(int j = i; j > i - k; --j)
            {
                currMax = max(currMax, arr[j]);
                dp[i] = max(dp[i], currMax * (i - j + 1) + dp[j-1]);
            }
        }

        return dp.back();
    }
};


int main()
{
    vector<int> test1 = {1,15,7,9,2,5,10};
    vector<int> test2 = {1,4,1,5,7,3,6,1,9,9,3};
    vector<int> test3 = {1};


    cout << Solution().maxSumAfterPartitioning(test1, 3) << endl;
    cout << Solution().maxSumAfterPartitioning(test2, 4) << endl;
    cout << Solution().maxSumAfterPartitioning(test3, 1) << endl;
}

/*
 Main idea:
 Пояснення з гарною візуалізацією цього алгоритму
 https://leetcode.com/problems/partition-array-for-maximum-sum/solutions/1621079/python-easy-dp-with-visualization-and-examples/?envType=daily-question&envId=2024-02-03

 Часова складість алгоритму: O(nk)
*/

File: DCC2024_Artur/February/Day3_Partition_Array_for_Maximum_Sum.py
"""
https://leetcode.com/problems/partition-array-for-maximum-sum/description/?envType=daily-question&envId=2024-02-03
"""

from typing import List

class Solution:
    def maxSumAfterPartitioning(self, arr: List[int], k: int) -> int:
        dp = [0 for _ in range(len(arr))]
        
        curr_max = 0
        for i in range(k):
            curr_max = max(curr_max, arr[i])
            dp[i] = curr_max * (i + 1)
        
        for i in range(k, len(arr)):
            curr_max = arr[i]
            for j in range(i, i -k, - 1):
               curr_max = max(curr_max, arr[j])
               dp[i] = max(dp[i], curr_max * (i - j + 1) + dp[j - 1])
        
        return dp[-1]
        

if __name__ == '__main__':
    
    print(Solution().maxSumAfterPartitioning([1,15,7,9,2,5,10], 3))
    print(Solution().maxSumAfterPartitioning([1,4,1,5,7,3,6,1,9,9,3], 4))
    print(Solution().maxSumAfterPartitioning([1], 1))
